### PR TITLE
[mono] Retarget HelloWorld sample to in-tree runtime pack (it used to use ".dotnet-mono" hack)

### LINKED ISSUE #40853: [mono] make run-tests-corefx-% target no longer works correctly
Running the following:

```
./build.sh mono+libs -c release
cd src/mono/netcore
make run-tests-corefx-System.Runtime.Loader
```

Causes us to loop on a particular line in mono.proj, logs below:

```
ryan@kenshin:~/git/runtime/src/mono/netcore$ make run-tests-corefx-System.Runtime
/Users/ryan/git/runtime/.dotnet/dotnet msbuild /t:RunLibrariesTest /p:LibraryToTest=System.Runtime /p:CoreClrTestConfig=Release /p:LibrariesTestConfig=Release /p:Configuration=Release ../mono.proj
Microsoft (R) Build Engine version 16.7.0-preview-20360-03+188921e2f for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Copying Mono Runtime artifacts from '/Users/ryan/git/runtime/artifacts/bin/mono/OSX.x64.Release/' to '/Users/ryan/git/runtime/artifacts/bin/testhost/net5.0-OSX-Release-x64/shared/Microsoft.NETCore.App/5.0.0/'.'
/Users/ryan/git/runtime/src/mono/mono.proj(851,5): warning MSB3026: Could not copy "/Users/ryan/git/runtime/artifacts/bin/mono/OSX.x64.Release/System.Private.CoreLib.dll" to "/Users/ryan/git/runtime/artifacts/bin/testhost/net5.0-OSX-Release-x64/shared/Microsoft.NETCore.App/5.0.0/System.Private.CoreLib.dll". Beginning retry 1 in 1000ms. Resource busy 
/Users/ryan/git/runtime/src/mono/mono.proj(851,5): warning MSB3026: Could not copy "/Users/ryan/git/runtime/artifacts/bin/mono/OSX.x64.Release/System.Private.CoreLib.pdb" to "/Users/ryan/git/runtime/artifacts/bin/testhost/net5.0-OSX-Release-x64/shared/Microsoft.NETCore.App/5.0.0/System.Private.CoreLib.pdb". Beginning retry 1 in 1000ms. Resource busy 
/Users/ryan/git/runtime/src/mono/mono.proj(851,5): warning MSB3026: Could not copy "/Users/ryan/git/runtime/artifacts/bin/mono/OSX.x64.Release/libcoreclr.dylib" to "/Users/ryan/git/runtime/artifacts/bin/testhost/net5.0-OSX-Release-x64/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.dylib". Beginning retry 1 in 1000ms. Resource busy 
Attempting to cancel the build...
```

Checking locally, it appears to have been copied. Not sure what's going on, but it would be nice to have that target working again.

### PR DESCRIPTION
**Pull request by EgorBo**
Also, get rid of various ancient ".dotnet-mono" hacks.

https://github.com/dotnet/runtime/pull/42597 was a blocker for this.

Also, fixes https://github.com/dotnet/runtime/issues/40853

### CODE REVIEW

File: src/mono/mono.proj [hunk 1]
```diff
@@ -810,52 +808,6 @@
...
-  <!-- Precompile (AssembliesToPrecompile) using (MonoExec) -->
-  <Target Name="PrecompileAssembly">
-   <PropertyGroup>
-      <EnvVars Condition="'$(OS)' == 'Windows_NT'">set MONO_ENV_OPTIONS=--aot=mcpu=native &amp; $(EnvVars)</EnvVars>
-      <EnvVars Condition="'$(OS)' != 'Windows_NT'">export MONO_ENV_OPTIONS=--aot=mcpu=native &amp;&amp; $(EnvVars)</EnvVars>
-    </PropertyGroup>
-    <Exec Command="$(EnvVars) $(MonoExec) %(AssembliesToPrecompile.Identity)" ContinueOnError="WarnAndContinue" />
-  </Target>
-
-  <!-- Precompile (AOT) managed libs in .dotnet-mono\shared\Microsoft.NETCore.App\**\*.dll -->
```
Comment (EgorBo) [inline src/mono/mono.proj:822]: These targets are unused. And we're going to use AOTCompiler task for it.

File: src/mono/netcore/Makefile [hunk 2]
```diff
@@ -27,14 +27,12 @@ MONO_PROJ=/p:CoreClrTestConfig=$(CORECLR_TESTS_CONFIG) /p:LibrariesTestConfig=$(
 
 # run sample using local .dotnet-mono
 # build it with .dotnet first in order to be able to use LLVM only for the actual sample
-run-sample: patch-mono-dotnet
-	$(DOTNET) build -c Release sample/HelloWorld
-	MONO_ENV_OPTIONS="--llvm" COMPlus_DebugWriteToStdErr=1 \
-	$(DOTNET_MONO) sample/HelloWorld/bin/HelloWorld.dll
+run-sample:
+	cd sample/HelloWorld && make run
```
Comment (CoffeeFlux) [inline src/mono/netcore/Makefile:31]: Shouldn't you just use `make -C` here?

File: src/mono/netcore/Makefile [hunk 3]
```diff
@@ -71,17 +61,3 @@ run-tests-coreclr-all:
 # for `System.Runtime.Tests` (the rule builds both the lib and the test if needed)
 run-tests-corefx-%:
 	$(DOTNET) msbuild /t:RunLibrariesTest /p:LibraryToTest=$* $(MONO_PROJ)
```
Comment (CoffeeFlux) [inline src/mono/netcore/Makefile:63]: Out of curiosity, does this work now with your changes?

File: src/mono/netcore/sample/HelloWorld/Makefile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	TARGET_OS=osx
+else
+	TARGET_OS=linux
+endif
+
+publish:
+	$(DOTNET) publish -c $(MONO_CONFIG) -r $(TARGET_OS)-$(MONO_ARCH)
+
+run: publish
+	MONO_ENV_OPTIONS="--llvm" \
```
Comment (CoffeeFlux) [inline src/mono/netcore/sample/HelloWorld/Makefile:19]: Please continue to set `COMPlus_DebugWriteToStdErr=1`, unless that's getting set somewhere else and I missed it.

File: src/mono/netcore/sample/HelloWorld/Program.cs [hunk 5]
```diff
@@ -7,10 +10,9 @@ internal class Program
         private static void Main(string[] args)
         {
             bool isMono = typeof(object).Assembly.GetType("Mono.RuntimeStructs") != null;
+            Console.WriteLine();
             Console.WriteLine("Hello World " + (isMono ? "from Mono!" : "from CoreCLR!"));
-            Console.WriteLine(typeof(object).Assembly.FullName);
-            Console.WriteLine(System.Reflection.Assembly.GetEntryAssembly ());
-            Console.WriteLine(System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription);
```
Comment (CoffeeFlux) [inline src/mono/netcore/sample/HelloWorld/Program.cs:13]: What's the motivation behind getting rid of this?

Comment (EgorBo) [reply src/mono/netcore/sample/HelloWorld/Program.cs:13]: ah, actually an unnecessary change

File: src/mono/netcore/Makefile (hunk 3 above)
Comment (EgorBo) [reply src/mono/netcore/Makefile:63]: Should work, the targets to patch testhost with mono bits are unchanged.
I'll check.

Comment (CoffeeFlux) [reply src/mono/netcore/Makefile:63]: If the testhost patching is unchanged it's probably still broken, just wanted to check.
Relevant issue: https://github.com/dotnet/runtime/issues/40853

Comment (EgorBo) [reply src/mono/netcore/Makefile:63]: ah, same output here.

**Review by CoffeeFlux — APPROVED** [review]
LGTM, thanks!

File: src/mono/netcore/Makefile (hunk 3 above)
Comment (EgorBo) [reply src/mono/netcore/Makefile]: Fixed, should work now

File: src/mono/mono.proj [hunk 6]
```diff
@@ -869,63 +817,36 @@
...
-
   <!-- Copy Mono runtime bits to the coreclr's Core_Root in order to run runtime tests -->
-  <Target Name="PatchCoreClrCoreRoot" DependsOnTargets="ValidateLocalDotnet">
+  <Target Name="PatchCoreClrCoreRoot">
     <MSBuild Projects="$(MSBuildProjectFullPath)"
              Properties="Destination=$(CoreClrTestCoreRoot)"
              Targets="CopyMonoRuntimeFilesFromArtifactsToDestination" />
   </Target>
 
   <!-- Copy Mono runtime bits to the coreclr's Core_Root in order to run runtime tests -->
```
Comment (lambdageek) [inline src/mono/mono.proj:827]: nit: this comment is wrong.

File: src/mono/netcore/sample/HelloWorld/Makefile [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+else
+	TARGET_OS=linux
+endif
+
+publish:
+	$(DOTNET) publish -c $(MONO_CONFIG) -r $(TARGET_OS)-$(MONO_ARCH)
+
+run: publish
+	COMPlus_DebugWriteToStdErr=1 \
+	MONO_ENV_OPTIONS="--llvm" \
```
Comment (lambdageek) [inline src/mono/netcore/sample/HelloWorld/Makefile:20]: Why is this `--llvm` here? can I run hello world with another EE?

File: src/mono/netcore/Makefile [hunk 8]
```diff
@@ -70,18 +58,6 @@ run-tests-coreclr-all:
...
-
-# precompile (AOT) all libraries' tests
-aot-tests-corefx:
-	$(DOTNET) msbuild /t:PrecompileLibrariesTests $(MONO_PROJ)
-
-# precompile (AOT) all libraries inside testhost (it's used for libraries' tests)
-aot-testhost-corefx:
-	$(DOTNET) msbuild /t:PrecompileTesthostLibraries $(MONO_PROJ)
+	$(DOTNET) msbuild /t:PatchLibrariesTesthost $(MONO_PROJ)
+	$(DOTNET) build ../../../src/libraries/$*/tests -c $(LIBRARIES_TESTS_CONFIG) /p:RuntimeFlavor=mono /t:Test
```
Comment (lambdageek) [inline src/mono/netcore/Makefile:62]: Why isn't this just `$(TOP)/build.sh -s mono+libs+libs.tests  -rc $(RUNTIME_CONFIG) -c $(LIBRARIES_TESTS_CONFIG) /p:RuntimeFlavor=mono /t:Test` or something like that?  I thought we didn't need to do patching by hand anymore

**Review by lambdageek — COMMENTED** [review]
I think this is a good start.   We should get rid of `src/mono/netcore/Makefile` and document  in `docs/workflow` how to do everything with `build.sh` and `dotnet.sh`

File: src/mono/netcore/sample/HelloWorld/Makefile (hunk 7 above)
Comment (EgorBo) [reply src/mono/netcore/sample/HelloWorld/Makefile:20]: Just a placeholder for custom args 🙂 if llvm is not built - it'll print a warning but the app will still work.

File: src/mono/netcore/Makefile (hunk 8 above)
Comment (steveisok) [reply src/mono/netcore/Makefile:62]: The desktop build should both generate an in-tree runtime pack and a testhost.  Unless I'm missing something, the `PatchLibrariesTesthost` target shouldn't be necessary.

**Comment by EgorBo** [comment]
> I think this is a good start. We should get rid of `src/mono/netcore/Makefile` and document in `docs/workflow` how to do everything with `build.sh` and `dotnet.sh`

I agree, just didn't want to break anybody's workflow in this PR.

File: src/mono/netcore/Makefile (hunk 8 above)
Comment (EgorBo) [reply src/mono/netcore/Makefile:62]: Yeah makes sense to drop it.

File: src/mono/mono.proj (hunk 6 above)
Comment (EgorBo) [reply src/mono/mono.proj:827]: hm.. looks correct to me 🤔

**Comment by CoffeeFlux** [comment]
> We should get rid of src/mono/netcore/Makefile and document in docs/workflow how to do everything with build.sh and dotnet.sh

I feel quite strongly in the opposite direction. The `build.sh` commands tend to be fairly verbose because they lack useful context, and from what I hear everyone solves this by copy-pasting snippets over and over. Our documentation should focus on that workflow for external contributors, but the makefile targets have additional context by virtue of living in the Mono tree that make them a much more streamlined experience when working on Mono. If this goes away I'll essentially be forced to recreate it in my local tree, which is what I had for wasm before some of the recent improvements.

**Comment by EgorBo** [comment]
So can we merge this?
I am going to add a few notes to the docs to cover some tips/tricks

**Review by marek-safar — APPROVED** [review]

**Comment by marek-safar** [comment]
@CoffeeFlux this PR does not remove src/mono/netcore/Makefile so merging it

**Pull request merged**